src/nodes.coffee

```diff
@@ -268,41 +272,101 @@ exports.Base = class Base
   # Plain JavaScript object representation of the node, that can be serialized
   # as JSON. This is used for generating an abstract syntax tree (AST).
   # This is what the `ast` option in the Node API returns.
-  toJSON: ->
+  toJSON: (o) ->
     # We try to follow the [Babel AST spec](https://github.com/babel/babel/blob/master/packages/babylon/ast/spec.md)
     # as closely as possible, for improved interoperability with other tools.
-    obj =
-      type: @constructor.name
-      # Convert `locationData` to Babel’s style.
-      loc:
-        start:
-          line: @locationData.first_line
-          column: @locationData.first_column
-        end:
-          line: @locationData.last_line
-          column: @locationData.last_column
-
-    # Add serializable properties to the output. Properties that aren’t
-    # automatically serializable (because they’re already a primitive type)
-    # should be handled on a case-by-case basis in child node classes’ own
-    # `toJSON` methods.
-    for property, value of this
-      continue if property in ['locationData', 'children']
-      continue if value is undefined # Don’t skip `null` or `false` values.
-      if typeof value is 'boolean' or typeof value is 'number' or typeof value is 'string'
-        obj[property] = value
-
-    # Work our way down the tree. This is like `eachChild`, except that we
-    # preserve the child node name, and arrays.
-    for attr in @children when @[attr]
-      if Array.isArray(@[attr])
-        obj[attr] = []
-        for child in flatten [@[attr]]
-          obj[attr].push child.unwrap().toJSON()
-      else
-        obj[attr] = @[attr].unwrap().toJSON()
+    @toAst o
+
+  # Returns the full AST for the node.
+  toAst: (o, level) ->
+    o = extend {}, o
+    o.level = level if level
+    @withBabylonLocationData @withAstType @_toAst(o), o
+
+  # Adds `type` to an AST.
+  # A node class typically defines `astType` (as a string or callback) if it
+  # needs to override the default-generated `type` (ie its constructor name)
+  withAstType: (ast, o) ->
+    return ast unless ast
+    return ast if isArray ast
+    return ast if ast.type
+    return ast unless @emptyAst or do ->
+      return yes for key in Object.keys(ast) when key not in ['comments', babylonLocationFields...]
+
+    merge ast,
+      type: do =>
+        return @constructor.name unless @astType
+        @astType?(o) ? @astType
+
+  # Returns the "content" (ie not `type` or location data) of the AST for the node.
+  # By default, recursively generates AST nodes for `children`.
+  # To override, a node class can either:
+  #   - define `astChildren` to specify how to generate full child nodes, and/or
+  #     `astProps` to specify how to generate non-child-node fields
+  #   - override the `_toAst()` method
+  _toAst: (o) ->
```

Comment:
Rather than `toJSON` that calls `toAst` that calls `_toAst`, a structure that would more closely follow the architecture of the nodes would be to have the logic that creates the properties that make up the AST split out as separate methods that may or not be overridden. So the base class would have a `type` method, for example, that generically defines the type of any node; and nodes for which that value doesn’t work can define their own `type` methods to override the base one. Then the base `toJSON` method (or whatever is assembling the AST) can just call `type`, and it’ll call either the base or overridden version. It’s better to split out methods by topic, like you’re describing with child nodes and simple fields, rather than “levels” like `toAst` and `_toAst`.

Reply:
@GeoffreyBooth ok first let's look at `type` as an example
> So the base class would have a `type` method, for example, that generically defines the type of any node; and nodes for which that value doesn’t work can define their own `type` methods to override the base one

So what you're describing is relatively close to what I have here. What you're describing as the `type()` method is basically what I've exposed as the `astType` method (or simple string value). I was encoding the default of `@constructor.name` into `withAstType()` (which is what's responsible for assembling the `type` field on the AST). But I like your approach better where the `Base` class provides the default, so I've updated and put `astType: -> @constructor.name` in `Base`

As far as the name `type()` vs `astType`, I think we want to keep AST-related stuff visibly identifiable as "AST"-related at least for now. I think `type()` is too generic of a name, for now at least we're better served by thinking of AST generation as a *transformation* step to generate a *new representation* rather than trying to conflate our node classes themselves with their AST representations

Then beyond `type`:
> a structure that would more closely follow the architecture of the nodes would be to have the logic that creates the properties that make up the AST split out as separate methods that may or not be overridden

I think this thinking only really works for `type` since it's the only "known" AST field (ie which has a known name that the AST generator could know to look for on the node class) besides location data. So then I was sort of taking inspiration from the way that `children` provides an elegantly declarative way of specifying how to recursively traverse the nodes tree. Thus the declarative "by topic" interfaces exposed by `astChildren`, `astProps` and `astType`
> rather than “levels” like `toAst` and `_toAst`

I do think there's a place for these "levels" as well. I've explained above how `toJSON()` shouldn't even really be seen as a level from the perspective of inside `nodes.coffee`, it just immediately hands off to the "internal" interfaces and is not something that anything inside `nodes.coffee` needs to be aware of

So then the only non-topic-based "levels" are `toAst()` and `_toAst()`. In practice, I found this to be a very useful split for maximizing reuse, as really what `toAst()` is responsible for is just the "known" fields (`type` and location data) that all AST nodes will have. So almost all node classes will want to get this "for free", looking at my `prettier` branch it's basically only a few node classes which don't want to be "seen as themselves" (but instead as eg a child node or an alternate representation of themselves) that override `toAst()`

Whereas `_toAst()`, whose responsibility is "everything besides those known fields", is much more likely to be overridden. Most node classes don't even need to override it and can just use the topic-based hooks `astChildren`/`astProps`/`astType`, but for a fair number of them it's more convenient to just "fully override" at the `_toAst()` level

Reply:
Maybe what’s needed is to make the overall naming more clear about what methods are intending to do. Perhaps `compileNode` should become `toJavaScriptString`, and the various `compile*` methods renamed accordingly (`toJavaScriptStringFragments`, etc.). That would also make it obvious which methods to remove once we have the printer up and running.

Along those lines, we can have `toAst` (or perhaps better would be `toAstJSON`) but we need a new name for `_toAst`, not least because we don’t have `_`-prefixed methods anywhere else in the codebase. What is that method doing that’s different than `toAst`? Can it be split up?

Sorry if it seems like I’m splitting hairs on bikeshed stuff, but I’m trying to understand the overall architecture of what you’re doing. It still feels wrong to me that we have three methods here (`toJSON`, `toAst`, `_toAst`). I want to at least understand the intent first, so that we can discuss if there’s a better way to organize the code.